File: codex/app/sources/semantic_scholar.py
from app.config import settings
from app.sources.base import SourceContext
from app.utils import is_title_match, safe_title_query


async def fetch_semantic_scholar(ctx: SourceContext, doi: str | None, title: str | None) -> dict | None:
    headers = {"User-Agent": settings.user_agent}
    if settings.semantic_scholar_api_key:
        headers["x-api-key"] = settings.semantic_scholar_api_key

    if doi:
        url = (
            "https://api.semanticscholar.org/graph/v1/paper/"
            f"DOI:{doi}?fields=title,openAccessPdf,externalIds,url"
        )
        data = await ctx.get_json(url, headers=headers)
        oa = data.get("openAccessPdf")
        if oa and oa.get("url"):
            return {"pdf_url": oa.get("url"), "metadata": data, "source": "semantic_scholar"}
        doi_value = (data.get("externalIds") or {}).get("DOI")
        if doi_value:
            return {
                "pdf_url": None,
                "doi": doi_value,
                "metadata": data,
                "source": "semantic_scholar",
            }
        return None

    if not title:
        return None

    url = (
        "https://api.semanticscholar.org/graph/v1/paper/search?query="
        f"{safe_title_query(title)}&limit=5&fields=title,openAccessPdf,externalIds,url"
    )
    data = await ctx.get_json(url, headers=headers)
    hits = data.get("data") or []
    if not hits:
        return None
    for hit in hits:
        if title and not is_title_match(title, hit.get("title", "")):
            continue
        oa = hit.get("openAccessPdf")
        if oa and oa.get("url"):
            return {"pdf_url": oa.get("url"), "metadata": hit, "source": "semantic_scholar"}
    for hit in hits:
        if title and not is_title_match(title, hit.get("title", "")):
            continue
        doi_value = (hit.get("externalIds") or {}).get("DOI")
        if doi_value:
            return {
                "pdf_url": None,
                "doi": doi_value,
                "metadata": hit,
                "source": "semantic_scholar",
            }
    return None

File: src/fetchers/semanticScholar.mjs
import axios from 'axios';
import { isTitleMatch } from '../utils/titleMatch.mjs';
import { withRetry } from './baseFetcher.mjs';

/**
 * Semantic Scholar fetcher - uses S2 API to find papers and their open access PDFs
 * Validates title similarity to avoid false positives
 * Now with retry logic for transient failures
 */
export async function fetchFromSemanticScholar(title) {
  try {
    const searchUrl = 'https://api.semanticscholar.org/graph/v1/paper/search';
    const params = {
      query: title,
      limit: 5,
      fields: 'title,authors,year,openAccessPdf,externalIds,url'
    };

    const headers = {};
    if (process.env.SEMANTIC_SCHOLAR_API_KEY) {
      headers['x-api-key'] = process.env.SEMANTIC_SCHOLAR_API_KEY;
    }

    const response = await withRetry(() =>
      axios.get(searchUrl, { params, headers, timeout: 10000 })
    );

    if (!response.data.data || response.data.data.length === 0) {
      return { success: false, error: 'No results found on Semantic Scholar' };
    }

    // Try to find a paper with matching title AND open access PDF
    for (const paper of response.data.data) {
      // Validate title similarity first
      if (!isTitleMatch(title, paper.title)) {
        continue;
      }

      if (paper.openAccessPdf?.url) {
        return {
          success: true,
          pdf_url: paper.openAccessPdf.url,
          source: 'Semantic Scholar',
          metadata: {
            title: paper.title,
            authors: paper.authors?.map(a => a.name).join(', '),
            year: paper.year,
            doi: paper.externalIds?.DOI,
            semanticScholarUrl: paper.url
          }
        };
      }
    }

    // Return DOI if found for matching title (for Unpaywall fallback)
    for (const paper of response.data.data) {
      if (isTitleMatch(title, paper.title) && paper.externalIds?.DOI) {
        return {
          success: false,
          error: 'No open access PDF on Semantic Scholar',
          doi: paper.externalIds.DOI
        };
      }
    }

    return { success: false, error: 'No matching paper found on Semantic Scholar' };
  } catch (error) {
    return { success: false, error: `Semantic Scholar error: ${error.message}` };
  }
}
